README.md
# ktor-rate-limit

A rate-limiting feature for Ktor servers.

```kotlin
install(RateLimit)

routing {
    rateLimited {
        // Rate limits apply to these routes
        get("/limited") {
            // ...            
        }       
    }
}
```

This library's implementation is loosely inspired by [mantono/ktor-rate-limiting](https://github.com/mantono/ktor-rate-limiting) but aims to be more reliable and have less code weirdnesses. It is *not* a fork, rather a new library that implements some of the good ideas of the original library.

This feature was originally implemented in [EpiLink](https://github.com/EpiLink/EpiLink) but has since been separated
into its own thing. `ktor-rate-limit`'s original code was under the MPL2 (no copyleft exception), but is available under
the Apache 2.0 license in this repository.

## Documentation

Documentation is on zgdocs [here](https://docs.zoroark.guru/#/ktor-rate-limit/). You can also check the KDoc comments,
which have a lot of information.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.0.2

* Bumped dependencies' versions
* Enabled explicit API mode internally.

## 0.0.1

* Added core rate-limiting functionality


src/test/kotlin/guru/zoroark/ratelimit/Tests.kt
/**
 * Copyright 2020 the original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package guru.zoroark.ratelimit

import io.ktor.application.ApplicationCall
import io.ktor.application.call
import io.ktor.application.install
import io.ktor.features.XForwardedHeaderSupport
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.response.respond
import io.ktor.routing.get
import io.ktor.routing.routing
import io.ktor.server.testing.TestApplicationCall
import io.ktor.server.testing.handleRequest
import io.ktor.server.testing.withTestApplication
import java.time.Duration
import java.time.Instant
import kotlin.math.ceil
import kotlin.test.*

class RateLimitTest {
    @Test
    fun `Test rate limiting headers`(): Unit = withTestApplication {
        val resetDur = Duration.ofMinutes(1)
        with(application) {
            install(RateLimit) {
                limit = 5
                timeBeforeReset = resetDur
            }
            routing {
                rateLimited {
                    get("/") {
                        call.respond(HttpStatusCode.OK)
                    }
                }
            }
        }
        // Add a second to account for possible lag or delays
        val max = Instant.now() + resetDur + Duration.ofSeconds(1)
        repeat(5) { iteration ->
            handleRequest(HttpMethod.Get, "/").apply {
                assertRateLimitedHeaders(5, 4L - iteration, max)
                assertStatus(HttpStatusCode.OK)
            }
        }

        handleRequest(HttpMethod.Get, "/").apply {
            assertStatus(HttpStatusCode.TooManyRequests)
            assertRateLimitedHeaders(5, 0, max)
            assertEquals(
                response.headers["X-RateLimit-Reset-After"]!!,
                response.headers["Retry-After"]
            )
        }
    }

    @Test
    fun `Different buckets have different counters`(): Unit =
        withTestApplication {
            val resetDur = Duration.ofMinutes(1)
            var useMe = "One"
            with(application) {
                install(RateLimit) {
                    limit = 5
                    timeBeforeReset = resetDur
                    callerKeyProducer = {
                        useMe.toByteArray()
                    }
                }
                routing {
                    rateLimited {
                        get("/") {
                            call.respond(HttpStatusCode.OK)
                        }
                    }
                }
            }
            // Add a second to account for possible lag or delays
            val max = Instant.now() + resetDur + Duration.ofSeconds(1)

            // Simulate first client
            repeat(4) { iteration ->
                handleRequest(HttpMethod.Get, "/").apply {
                    assertRateLimitedHeaders(5, 4L - iteration, max)
                    assertStatus(HttpStatusCode.OK)
                }
            }
            // Simulate second client
            useMe = "Two"
            repeat(3) { iteration ->
                handleRequest(HttpMethod.Get, "/").apply {
                    assertRateLimitedHeaders(5, 4L - iteration, max)
                }
            }
            // Simulate first client, whose bucket should not have expired
            useMe = "One"
            handleRequest(HttpMethod.Get, "/").apply {
                assertRateLimitedHeaders(5, 0L, max)
                assertStatus(HttpStatusCode.OK)
            }
            handleRequest(HttpMethod.Get, "/").apply {
                assertStatus(HttpStatusCode.TooManyRequests)
                assertRateLimitedHeaders(5, 0, max)
                assertEquals(
                    response.headers["X-RateLimit-Reset-After"]!!,
                    response.headers["Retry-After"]
                )
            }
        }

    @Test
    fun `Different routes have different counters`(): Unit =
        withTestApplication {
            val resetDur = Duration.ofMinutes(1)
            with(application) {
                install(RateLimit) {
                    limit = 5
                    timeBeforeReset = resetDur
                }
                routing {
                    rateLimited {
                        get("/one") {
                            call.respond(HttpStatusCode.OK)
                        }
                    }
                    rateLimited {
                        get("/two") {
                            call.respond(HttpStatusCode.OK)
                        }
                    }
                }
            }
            // Add a second to account for possible lag or delays
            val max = Instant.now() + resetDur + Duration.ofSeconds(1)

            // Simulate first route
            repeat(4) { iteration ->
                handleRequest(HttpMethod.Get, "/one").apply {
                    assertRateLimitedHeaders(5, 4L - iteration, max)
                    assertStatus(HttpStatusCode.OK)
                }
            }
            // Simulate second route
            repeat(3) { iteration ->
                handleRequest(HttpMethod.Get, "/two").apply {
                    assertRateLimitedHeaders(5, 4L - iteration, max)
                }
            }
            // Simulate second route, whose bucket should not have expired
            handleRequest(HttpMethod.Get, "/one").apply {
                assertRateLimitedHeaders(5, 0L, max)
                assertStatus(HttpStatusCode.OK)
            }
            handleRequest(HttpMethod.Get, "/one").apply {
                assertStatus(HttpStatusCode.TooManyRequests)
                assertRateLimitedHeaders(5, 0, max)
                assertEquals(
                    response.headers["X-RateLimit-Reset-After"]!!,
                    response.headers["Retry-After"]
                )
            }
        }

    @Test
    fun `Different additional key have different counters`(): Unit =
        withTestApplication {
            val resetDur = Duration.ofMinutes(1)
            with(application) {
                install(RateLimit) {
                    limit = 5
                    timeBeforeReset = resetDur
                }
                routing {
                    rateLimited(additionalKeyExtractor = {
                        parameters["id"]!!
                    }) {
                        get("/{id}") {
                            call.respond(HttpStatusCode.OK)
                        }
                    }
                }
            }
            // Add a second to account for possible lag or delays
            val max = Instant.now() + resetDur + Duration.ofSeconds(1)

            // Simulate first key
            repeat(4) { iteration ->
                handleRequest(HttpMethod.Get, "/one").apply {
                    assertRateLimitedHeaders(5, 4L - iteration, max)
                    assertStatus(HttpStatusCode.OK)
                }
            }
            // Simulate second key
            repeat(3) { iteration ->
                handleRequest(HttpMethod.Get, "/two").apply {
                    assertRateLimitedHeaders(5, 4L - iteration, max)
                }
            }
            // Simulate second key, whose bucket should not have expired
            handleRequest(HttpMethod.Get, "/one").apply {
                assertRateLimitedHeaders(5, 0L, max)
                assertStatus(HttpStatusCode.OK)
            }
            handleRequest(HttpMethod.Get, "/one").apply {
                assertStatus(HttpStatusCode.TooManyRequests)
                assertRateLimitedHeaders(5, 0, max)
                assertEquals(
                    response.headers["X-RateLimit-Reset-After"]!!,
                    response.headers["Retry-After"]
                )
            }
        }

    @Test
    fun `Rate limit expires`(): Unit = withTestApplication {
        val resetDur = Duration.ofMillis(300)
        with(application) {
            install(RateLimit) {
                limit = 3
                timeBeforeReset = resetDur
            }
            routing {
                rateLimited(additionalKeyExtractor = {
                    parameters["id"]!!
                }) {
                    get("/{id}") {
                        call.respond(HttpStatusCode.OK)
                    }
                }
            }
        }
        repeat(2) {
            handleRequest(HttpMethod.Get, "/one").apply {
                assertStatus(HttpStatusCode.OK)
            }
        }
        repeat(3) {
            handleRequest(HttpMethod.Get, "/two").apply {
                assertStatus(HttpStatusCode.OK)
            }
        }
        handleRequest(HttpMethod.Get, "/two").apply {
            assertStatus(HttpStatusCode.TooManyRequests)
        }
        Thread.sleep(300)
        handleRequest(HttpMethod.Get, "/one").apply {
            assertStatus(HttpStatusCode.OK)
            assertEquals("2", response.headers["X-RateLimit-Remaining"])
        }
        handleRequest(HttpMethod.Get, "/two").apply {
            assertStatus(HttpStatusCode.OK)
            assertEquals("2", response.headers["X-RateLimit-Remaining"])
        }
    }

    // The following test cannot be ran automatically. Uncomment it and run it.
    // You should see in the DEBUG logs a lot of output from
    // guru.zoroark.ratelimit.inmemory like "Removing stale bucket abcdefg...=":
    // that means the purge has ran.
    /*
    @Test
    fun `Check purge has run`(): Unit = withTestApplication {
        val resetDur = Duration.ofMillis(300)
        with(application) {
            install(RateLimit) {
                limit = 3
                timeBeforeReset = resetDur
                limiter = InMemoryRateLimiter(5, Duration.ofMillis(400))
            }
            routing {
                rateLimited(additionalKeyExtractor = {
                    parameters["id"]!!
                }) {
                    get("/{id}") {
                        call.respond(HttpStatusCode.OK)
                    }
                }
            }
        }
        repeat(100) {
            handleRequest(HttpMethod.Get, "/$it").apply {
                assertStatus(HttpStatusCode.OK)
            }
        }
        Thread.sleep(400)
        handleRequest(HttpMethod.Get, "/1").apply {
            assertStatus(HttpStatusCode.OK)
        }
        Thread.sleep(1000) // Wait for the purge to happen, it runs in the background
    }
    */

    @Test
    fun `Test rate limiting same header on IPv6 same network prefix`() {
        withTestApplication {
            with(application) {
                install(RateLimit)
                routing {
                    rateLimited {
                        get("/") {
                            call.respond(HttpStatusCode.OK)
                        }
                    }
                }

                install(XForwardedHeaderSupport)
            }
            val response1 = handleRequest(HttpMethod.Get, "/") {
                addHeader(
                    HttpHeaders.XForwardedFor,
                    "1c68:5440:2594:c1e6:bfdd:8b97:e61d:b2ff"
                )
            }
            val response2 = handleRequest(HttpMethod.Get, "/") {
                addHeader(
                    HttpHeaders.XForwardedFor,
                    "1c68:5440:2594:c1e6:efc5:7a53:7135:41f"
                )
            }
            assertEquals(
                response1.response.headers["X-RateLimit-Bucket"],
                response2.response.headers["X-RateLimit-Bucket"]
            )
        }
    }


    @Test
    fun `Test rate limiting same header on IPv6 different network prefix`() {
        withTestApplication {
            with(application) {
                install(RateLimit)
                routing {
                    rateLimited {
                        get("/") {
                            call.respond(HttpStatusCode.OK)
                        }
                    }
                }

                install(XForwardedHeaderSupport)
            }
            val response1 = handleRequest(HttpMethod.Get, "/") {
                addHeader(
                    HttpHeaders.XForwardedFor,
                    "1c68:5440:2594:c1e6:bfdd:8b97:e61d:b2ff"
                )
            }
            val response2 = handleRequest(HttpMethod.Get, "/") {
                addHeader(
                    HttpHeaders.XForwardedFor,
                    "7e92:2075:327e:f32b:4341:97a2:1440:5a2f"
                )
            }
            assertNotEquals(
                response1.response.headers["X-RateLimit-Bucket"],
                response2.response.headers["X-RateLimit-Bucket"]
            )
        }
    }
}

/*
 * Has some weirdness to avoid rounding errors
 */
fun TestApplicationCall.assertRateLimitedHeaders(
    expectedLimit: Long,
    expectedRemaining: Long,
    resetMax: Instant,
    inMillis: Boolean = false
) {
    val actualResetMax =
        if (inMillis)
            resetMax
        else
            Instant.ofEpochSecond(ceil(resetMax.toEpochMilli() / 1000.0).toLong())

    assertEquals(
        expectedLimit.toString(),
        response.headers["X-RateLimit-Limit"]
    )
    assertEquals(
        expectedRemaining.toString(),
        response.headers["X-RateLimit-Remaining"]
    )
    val resetAt =
        if (inMillis)
            Instant.ofEpochMilli((response.headers["X-RateLimit-Reset"]!!.toDouble() * 1000).toLong())
        else
            Instant.ofEpochSecond((response.headers["X-RateLimit-Reset"]!!.toLong()))
    response.headers["X-RateLimit-Reset"]!!.toLong()
    println("Check Reset at ${resetAt.toEpochMilli()} <= actual reset max ${actualResetMax.toEpochMilli()}")
    assertTrue(resetAt <= actualResetMax, "Resets before max")
    val resetAfter =
        if (inMillis)
            Duration.ofMillis((response.headers["X-RateLimit-Reset-After"]!!.toDouble() * 1000).toLong())
        else
            Duration.ofSeconds(response.headers["X-RateLimit-Reset-After"]!!.toLong())
    val expectedResetAfterMillis =
        Duration.between(Instant.now(), actualResetMax).toMillis()
    val expectedResetAfter =
        if (inMillis)
            Duration.ofMillis(expectedResetAfterMillis)
        else
            Duration.ofSeconds(ceil(expectedResetAfterMillis / 1000.0).toLong())
    println("Check reset after ${resetAfter.toMillis()} <= ${expectedResetAfter.toMillis()}}")
    assertTrue(
        resetAfter <= expectedResetAfter,
        "Reset after less than max"
    )
    assertNotNull(response.headers["X-RateLimit-Bucket"], "Has a bucket")
}

private fun ApplicationCall.assertStatus(code: HttpStatusCode) {
    assertEquals(code, this.response.status())
}

src/main/kotlin/guru/zoroark/ratelimit/Rate.kt
/**
 * Copyright 2020 the original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package guru.zoroark.ratelimit

import java.time.Instant

/**
 * Represents a single rate limit state. Immutable.
 */
public data class Rate(
    /**
     * The number of remaining requests + 1. So, if this is 1, then this is the last allowed request and the next
     * request will trigger a rate limit.
     */
    val remainingRequests: Long,
    /**
     * The instant at which the rate limit is invalid and reset.
     */
    val resetAt: Instant
)

/**
 * Returns true if the rate has expired (we have passed its reset instant)
 */
public fun Rate.hasExpired(): Boolean = resetAt < Instant.now()

/**
 * Consumes a single request and returns the modified rate.
 */
public fun Rate.consume(): Rate = copy(remainingRequests = (remainingRequests - 1).coerceAtLeast(0))

/**
 * Returns true if the 429 HTTP error should be returned, as in, this rate has not expired and there are no remaining
 * requests
 */
public fun Rate.shouldLimit(): Boolean = !hasExpired() && remainingRequests == 0L


src/main/kotlin/guru/zoroark/ratelimit/Utils.kt
/**
 * Copyright 2020 the original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package guru.zoroark.ratelimit

import io.ktor.application.ApplicationCall
import io.ktor.features.origin
import org.slf4j.Logger
import java.net.Inet6Address

/**
 * Utility function for logging a lazily-constructed debug message.
 */
internal inline fun Logger.debug(lazyMsg: () -> String) {
    if (this.isDebugEnabled) {
        debug(lazyMsg())
    }
}

/**
 * The default caller key producer.
 */
fun ApplicationCall.defaultCallerKeyProducer(): ByteArray {
    val host = request.origin.remoteHost
    // Try parsing this as an IPV6 address.
    runCatching {
        Inet6Address.getByName("[$host]")
    }.onSuccess {
        // This is an IPv6 address: only take the first 64 bits (8 bytes)
        // as the caller key
        return it.address.copyOf(8)
    }
    return host.toByteArray()
}

src/main/kotlin/guru/zoroark/ratelimit/RateLimitingContext.kt
/**
 * Copyright 2020 the original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package guru.zoroark.ratelimit

import java.time.Duration
import java.time.Instant

/**
 * This context provides information useful for [RateLimiter]s that wish to create new [Rate] objects.
 */
public data class RateLimitingContext(
    /**
     * The limit, which is the maximum number amount of requests that can be made in the span of [resetTime]
     * milliseconds
     */
    val limit: Long,
    /**
     * The amount of milliseconds the rate limit lasts
     */
    val resetTime: Long
)

/**
 * Creates a new [Rate] object based on the current rate limiting context. Useful for creating brand new rates.
 */
public fun RateLimitingContext.newRate(): Rate =
    Rate(limit, Instant.now() + Duration.ofMillis(resetTime))


src/main/kotlin/guru/zoroark/ratelimit/RateLimit.kt
/**
 * Copyright 2020 the original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package guru.zoroark.ratelimit

import guru.zoroark.ratelimit.RateLimit.Configuration
import io.ktor.application.*
import io.ktor.features.origin
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.ktor.request.ApplicationRequest
import io.ktor.request.header
import io.ktor.response.ApplicationResponse
import io.ktor.response.header
import io.ktor.response.respondText
import io.ktor.routing.*
import io.ktor.util.AttributeKey
import io.ktor.util.pipeline.PipelineContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import java.security.MessageDigest
import java.security.SecureRandom
import java.time.Duration
import java.time.Instant
import java.util.*
import kotlin.math.ceil

/**
 * This feature implements rate limiting functionality.
 *
 * Rate limiting is when an API will only allow a certain number of calls be
 * made during a certain amount of time. This feature implements this behavior.
 * Rate limiting can be done on a per-route basis, meaning that different routes
 * may have different rate limits and be timed differently.
 *
 * This rate-limiting functionality is close to the one
 * [implemented by Discord](https://discordapp.com/developers/docs/topics/rate-limits),
 * with the notable exception that the `Retry-After` header returns durations in
 * **seconds**, not milliseconds.
 * [Discord's implementation does not follow standards](https://github.com/discord/discord-api-docs/issues/1463).
 *
 * This feature by itself does not limit anything, you need to define which
 * routes are rate limited using the [rateLimited] function. For example:
 *
 * ```
 *  install(RateLimit)
 *  // ...
 *  routing {
 *      route("user") {
 *          rateLimited(...) {
 *              get("something") { ... }
 *              post("somethingElse") { ... }
 *          }
 *      }
 *  }
 * ```
 *
 * Each rate-limited route can have its own limits and reset times. Check the
 * [rateLimited] function for more information.
 *
 * Each rate limit has a unique identifier (called a bucket) that is made of:
 *
 * - A **caller key** unique to the requester, determined using the
 * [callerKeyProducer][Configuration.callerKeyProducer].
 *
 * - A **routing key** unique to the route (where the [rateLimited] function is
 * used). Randomly generated.
 *
 * - An **additional key**, which is especially useful if your route matches
 * multiple paths (e.g. there is some ID in your path) and you want each path to
 * have its own individual limit.
 *
 * All of these keys are SHA-1'd together and turned into a Base 64 string: that
 * is the bucket we return.
 *
 * When rate-limited, this feature will end the pipeline immediately, returning
 * a HTTP 429 error with a JSON object.
 *
 * ```
 *  {
 *      "message": "You are being rate limited.",
 *      "retry_after": 12345,
 *      "global": false
 *  }
 * ```
 *
 * `global` is always false (global rate limits are not implemented yet) and
 * `retry_after` has the same value as the `Retry-After` header
 *
 * Global rate limits are not implemented yet.
 */
public class RateLimit(configuration: Configuration) {

    /**
     * The (non-standard) headers used by the rate limiting features. The names
     * are identical to the ones used by Discord.
     */
    public object Headers {
        /**
         * The number of requests that can be made
         */
        public const val Limit: String = "X-RateLimit-Limit"

        /**
         * The number of remaining requests that can be made before the reset time.
         */
        public const val Remaining: String = "X-RateLimit-Remaining"

        /**
         * The epoch time at which the rate limit resets
         */
        public const val Reset: String = "X-RateLimit-Reset"

        /**
         * The total time in seconds (either integer or floating) of when the
         * current rate limit will reset
         */
        public const val ResetAfter: String = "X-RateLimit-Reset-After"

        /**
         * Header for precision (second or millisecond)
         */
        public const val Precision: String = "X-RateLimit-Precision"

        /**
         * "Bucket" header
         */
        public const val Bucket: String = "X-RateLimit-Bucket"
    }

    /**
     * Configuration class for the Rate Limiting feature
     */
    public class Configuration {
        /**
         * The limiter implementation. See [RateLimiter] for more information.
         *
         * The [RateLimiter] is intended for use with Strings here.
         *
         * Default: An [InMemoryRateLimiter] which needs 100 items stored to
         * purge itself (one purge per hour).
         */
        public var limiter: RateLimiter<String> =
            InMemoryRateLimiter(100, Duration.ofHours(1))

        /**
         * The default limit (i.e. amount of requests) allowed for per-route
         * rate limits. Can be overridden via the [rateLimited] function.
         *
         * Default: 50 requests
         */
        public var limit: Long = 50L

        /**
         * The default amount of time before a rate limit is reset. Can be
         * overridden via the [rateLimited] function.
         *
         * Default: 2 minutes
         */
        public var timeBeforeReset: Duration = Duration.ofMinutes(2)

        /**
         * This is the function that generates caller keys. The default uses the
         * remote host as the caller key.
         */
        public var callerKeyProducer: ApplicationCall.() -> ByteArray = {
            request.origin.remoteHost.toByteArray()
        }
    }

    internal val random = SecureRandom()
    private val rateLimiter = configuration.limiter
    private val limit = configuration.limit
    private val timeBeforeReset = configuration.timeBeforeReset.toMillis()
    private val keyProducer = configuration.callerKeyProducer
    private val logger = LoggerFactory.getLogger("guru.zoroark.ratelimit")

    /**
     * Feature companion object for the rate limiting feature
     */
    public companion object Feature :
        ApplicationFeature<ApplicationCallPipeline, Configuration, RateLimit> {
        override val key: AttributeKey<RateLimit> = AttributeKey("RateLimit")

        override fun install(
            pipeline: ApplicationCallPipeline,
            configure: Configuration.() -> Unit
        ): RateLimit {
            return RateLimit(Configuration().apply(configure))
        }
    }

    /**
     * Handles a rate-limited call.
     *
     * @param limit Limit override
     * @param timeBeforeReset Reset time override (in milliseconds)
     */
    internal suspend fun handleRateLimitedCall(
        limit: Long?,
        timeBeforeReset: Long?,
        context: PipelineContext<Unit, ApplicationCall>,
        fullKeyProcessor: suspend (ByteArray) -> String
    ) = with(context.call) {
        // Initialize values
        val bucket = fullKeyProcessor(context.call.keyProducer())
        val actualLimit = limit ?: this@RateLimit.limit
        val actualTimeBeforeReset =
            timeBeforeReset ?: this@RateLimit.timeBeforeReset
        val rateContext =
            RateLimitingContext(actualLimit, actualTimeBeforeReset)
        // Handle the rate limit
        val rate = rateLimiter.handle(rateContext, bucket)
        // Append information to reply
        val inMillis = request.shouldRateLimitTimeBeInMillis()
        val remainingTimeBeforeReset =
            Duration.between(Instant.now(), rate.resetAt)
        response.appendRateLimitHeaders(
            rate,
            inMillis,
            remainingTimeBeforeReset,
            rateContext,
            bucket
        )
        // Interrupt call if we should limit it
        if (rate.shouldLimit()) {
            logger.debug { "Bucket $bucket (remote host ${request.origin.remoteHost}) is being rate limited, resets at ${rate.resetAt}" }
            val retryAfter = toHeaderValueWithPrecision(
                false,
                remainingTimeBeforeReset.toMillis()
            )
            // Always in seconds
            response.header(HttpHeaders.RetryAfter, retryAfter)
            respondText(
                ContentType.Application.Json,
                HttpStatusCode.TooManyRequests
            ) {
                """{"message":"You are being rate limited.","retry_after":$retryAfter,"global":false}"""
            }
            context.finish()
        } else {
            logger.debug {
                "Bucket $bucket (remote host ${request.origin.remoteHost}) passes rate limit, remaining = ${rate.remainingRequests - 1}, resets at ${rate.resetAt}"
            }
            context.proceed()
        }
    }
}

/**
 * Intercepts every call made inside the route block and adds rate-limiting to it.
 *
 * This function requires the [RateLimit] feature to be installed.
 *
 * Optionally, you can override some parameters that will only apply to this route.
 *
 * @param limit Overrides the global limit set when configuring the feature.
 * Maximum amount of requests that can be performed before being rate-limited
 * and receiving HTTP 429 errors.
 * @param timeBeforeReset Overrides the global time before reset set when
 * configuring the feature. Time before a rate-limit expires.
 * @param additionalKeyExtractor Function used for retrieving the additional
 * key. See [RateLimit] for more information.
 * @param callback Block for configuring the rate-limited route
 */
public fun Route.rateLimited(
    limit: Long? = null,
    timeBeforeReset: Duration? = null,
    additionalKeyExtractor: ApplicationCall.() -> String = { "" },
    callback: Route.() -> Unit
): Route {
    // Create the route
    val rateLimitedRoute = createChild(object : RouteSelector() {
        override fun evaluate(
            context: RoutingResolveContext,
            segmentIndex: Int
        ) =
            RouteSelectorEvaluation.Constant
    })
    // Rate limiting feature object
    val rateLimiting = application.feature(RateLimit)
    // Generate a key for this route
    val arr = ByteArray(64)
    rateLimiting.random.nextBytes(arr)
    // Intercepting every call and checking the rate limit
    rateLimitedRoute.intercept(ApplicationCallPipeline.Features) {
        rateLimiting.handleRateLimitedCall(
            limit,
            timeBeforeReset?.toMillis(),
            this
        ) {
            // This is the key generation. We simply SHA1 together all three keys.
            sha1(it, call.additionalKeyExtractor().toByteArray(), arr)
        }
    }
    rateLimitedRoute.callback()
    return rateLimitedRoute
}

/**
 * Shortcut function for checking if the precision of the rate limit should be in milliseconds (true) or seconds (false)
 */
private fun ApplicationRequest.shouldRateLimitTimeBeInMillis(): Boolean =
    header(RateLimit.Headers.Precision) == "millisecond"

/**
 * Appends rate-limiting related headers to the response
 */
private fun ApplicationResponse.appendRateLimitHeaders(
    rate: Rate,
    inMillis: Boolean,
    remainingTimeBeforeReset: Duration,
    context: RateLimitingContext,
    bucket: String
) {
    header(RateLimit.Headers.Limit, context.limit)
    header(
        RateLimit.Headers.Remaining,
        (rate.remainingRequests - 1).coerceAtLeast(0)
    )
    header(
        RateLimit.Headers.Reset,
        toHeaderValueWithPrecision(inMillis, rate.resetAt.toEpochMilli())
    )
    header(
        RateLimit.Headers.ResetAfter,
        toHeaderValueWithPrecision(
            inMillis,
            remainingTimeBeforeReset.toMillis()
        )
    )
    header(RateLimit.Headers.Bucket, bucket)
}

/**
 * Turns [millis] into either seconds (integer) if [inMillis] is false or seconds with milliseconds precision (double)
 * if [inMillis] is true and convert them to a String.
 *
 * @param millis A value in milliseconds
 */
private fun toHeaderValueWithPrecision(
    inMillis: Boolean,
    millis: Long
): String {
    return if (inMillis)
        (millis / 1000.0).toString()
    else
        ceil(millis / 1000.0).toInt().toString()
}

/**
 * Creates a Base 64 string from SHA-1'ing all of the arrays, treating them as a single byte array
 */
private suspend fun sha1(vararg arr: ByteArray): String =
    withContext(Dispatchers.Default) {
        val md = MessageDigest.getInstance("SHA-1")
        arr.forEach { md.update(it) }
        Base64.getEncoder().encodeToString(md.digest())
    }
